src/python.rs

```diff
@@ -189,133 +211,135 @@ impl Python<'_> {
-    ///
-    /// **Note:** This return type from this function, `GILGuard`, is implemented as a RAII guard
-    /// around the C-API Python_EnsureGIL. This means that multiple `acquire_gil()` calls are
-    /// allowed, and will not deadlock. However, `GILGuard`s must be dropped in the reverse order
-    /// to acquisition. If PyO3 detects this order is not maintained, it may be forced to begin
-    /// an irrecoverable panic.
+    /// 1. If implementing [`#[pymethods]`](crate::pymethods) or [`#[pyfunction]`](crate::pyfunction),  declare `py: Python` as an argument.
+    /// PyO3 will pass in the token to grant access to the GIL context in which the function is running.
+    /// 2. Use [`Python::with_gil`] to run a closure with the GIL, acquiring only if needed.
     ///
     /// # Panics
-    /// - If the `auto-initialize` feature is not enabled and the Python interpreter is not
-    ///   initialized.
+    ///
+    /// - If the [`auto-initialize`] feature is not enabled and the Python interpreter is not
+    /// initialized.
+    /// - If multiple [`GILGuard`]s are not dropped in in the reverse order of acquisition, PyO3
+    /// may panic. It is recommended to use [`Python::with_gil`] instead to avoid this.
+    ///
+    /// # Notes
+    ///
+    /// The return type from this function, [`GILGuard`], is implemented as a RAII guard
+    /// around [`PyGILState_Ensure`]. This means that multiple `acquire_gil()` calls are
+    /// allowed, and will not deadlock. However, [`GILGuard`]s must be dropped in the reverse order
+    /// to acquisition. If PyO3 detects this order is not maintained, it may be forced to begin
+    /// a panic.
+    ///
+    /// [`PyGILState_Ensure`]: crate::ffi::PyGILState_Ensure
+    /// [`auto-initialize`]: https://pyo3.rs/main/features.html#auto-initialize
+    /// [`prepare_freethreaded_python`]: crate::prepare_freethreaded_python
     #[inline]
     pub fn acquire_gil() -> GILGuard {
         GILGuard::acquire()
     }
 
-    /// Temporarily releases the `GIL`, thus allowing other Python threads to run.
+    /// Temporarily releases the GIL, thus allowing other Python threads to run. The GIL will be
+    /// reacquired when `F`'s scope ends.
     ///
-    /// # Examples
+    /// If you don't need to touch the Python
+    /// interpreter for some time and have other Python threads around, this will let you run
+    /// Rust-only code while letting those other Python threads make progress.
+    ///
+    /// The closure is impermeable to types that are tied to holding the GIL, such as `&`[`PyAny`]
+    /// and its concrete-typed siblings like `&`[`PyString`]. This is achieved via the [`Send`]
+    /// bound on the closure and the return type. This is slightly
+    /// more restrictive than necessary, but it's the most fitting solution available in stable
+    /// Rust. In the future this bound may be relaxed by using an "auto-trait" instead, if
+    /// [auto-traits] ever become a stable feature of the Rust language.
```

Comment:
Maybe link to https://github.com/PyO3/pyo3/issues/538 here?

Reply:
I don't think a link to the issue really adds anything. It's not some sort of "needs-implementer" thing where someone can just contribute a fix if they need to get around it. And we're not getting auto-traits any time soon (if ever) so it's not something that people can look forward to getting fixed.